beacon-chain/core/blocks/block_operations.go

```diff
@@ -141,26 +130,41 @@ func ProcessBlockHeader(
 		return nil, fmt.Errorf("proposer at index %d was previously slashed", idx)
 	}
 	if verifySignatures {
-		pub, err := bls.PublicKeyFromBytes(proposer.Pubkey)
-		if err != nil {
-			return nil, fmt.Errorf("could not deserialize proposer public key: %v", err)
-		}
-		domain := helpers.Domain(beaconState, helpers.SlotToEpoch(block.Slot), params.BeaconConfig().DomainBeaconProposer)
-		sig, err := bls.SignatureFromBytes(block.Signature)
-		if err != nil {
-			return nil, fmt.Errorf("could not convert bytes to signature: %v", err)
-		}
-		root, err := ssz.SigningRoot(block)
-		if err != nil {
-			return nil, fmt.Errorf("could not sign root for header: %v", err)
-		}
-		if !sig.Verify(root[:], pub, domain) {
-			return nil, fmt.Errorf("block signature did not verify")
+		if err := verifyProposerSignature(beaconState, block, idx); err != nil {
+			return nil, fmt.Errorf("could not verify proposer signature: %v", err)
 		}
 	}
 	return beaconState, nil
 }
 
+// verifyProposerSignature uses BLS signature verification to ensure
+// the correct proposer created an incoming beacon block during state
+// transition processing.
+//
+// Spec pseudocode definition:
+//   assert bls_verify(proposer.pubkey, signing_root(block), block.signature, get_domain(state, DOMAIN_BEACON_PROPOSER))
+func verifyProposerSignature(beaconState *pb.BeaconState, block *pb.BeaconBlock, proposerIdx uint64) error {
```

Comment:
this seems like a very standard pattern, can we generalize it to just `verifySignature`?

we can use this in `ProcessTransfers`, `ProcessBlockHeader`, `ProcessAttestation` and more...

Reply:
Good idea!